Fix spotless cache miss

tylerbertrand opened the pull request:
# Description

Spotless custom steps are never up to date by default. This is achieved using a randomly generated number attached to the custom step's state, which causes the custom step to never be up-to-date between runs.

Running Gradle's [Build Validation Scripts](https://github.com/gradle/gradle-enterprise-build-validation-scripts#gradle-enterprise-build-validation-scripts) with the `spotlessCheck` task shows the [task input differences causing the spotless tasks to be re-executed unnecessarily](https://ge.solutions-team.gradle.com/c/mgjwt7or35s5o/abkb2i4bo57cc/task-inputs?cacheability=cacheable,overlapping-outputs,validation-failure)

# Solution

Calling `bumpThisNumberIfACustomStepChanges(int)` pins this randomly generated number to the provided int, but it must be updated if a custom step changes, otherwise incorrect cache hits are possible.
* [link to `bumpThisNumberIfACustomStepChanges(int)` source](https://github.com/diffplug/spotless/blob/gradle/6.5.2/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/FormatExtension.java#L324-L342) 

# Tests

1. With these changes applied, running Gradle's [Build Validation Scripts](https://github.com/gradle/gradle-enterprise-build-validation-scripts#gradle-enterprise-build-validation-scripts) with the `spotlessCheck` task shows that the [task input differences have been resolved](https://ge.solutions-team.gradle.com/c/trjb4xwrqrw52/25l6rrrw5s6xo/task-inputs?cacheability=cacheable,overlapping-outputs,validation-failure)

2. With these changes applied, a wildcard import was added, and `spotlessCheck` was executed. This correctly [failed the build](https://ge.solutions-team.gradle.com/s/lqq4qxqwkkbai/failure#1).

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check` ([build scan](https://ge.solutions-team.gradle.com/s/3qliiquklepa4)).
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

runningcode commented on gradle/documentation/render-javadoc.gradle:
```diff
@@ -61,7 +61,14 @@ allprojects {
-      taskResources = resources
+      taskResources = project.relativePath(resources)
+
+      gradleEnterprise {
```
I think we should remove this entire block.

Just for learning's sake: even if we were to keep this here, the `gradleEnterprise` line adds no value.

runningcode commented on gradle/java/folder-layout.gradle:
```diff
@@ -25,13 +25,11 @@ allprojects {
     }
 
-    task copyTestResources(type: Copy) {
+    processTestResources {
```
let's use this to avoid eager task configuration:

```suggestion
    tasks.named('processTestResources') {
```

runningcode commented on gradle/testing/defaults-tests.gradle:
```diff
@@ -23,6 +23,23 @@ import org.apache.lucene.gradle.ErrorReportingTestListener
 def resources = scriptResources(buildscript)
 def verboseModeHookInstalled = false
 
+class ArgumentProvider implements CommandLineArgumentProvider {
```
let's give this class a better name like `LoggingFileArgumentProvider`

can we put this at the end of the file?

runningcode commented on gradle/testing/randomization.gradle:
```diff
@@ -34,6 +34,28 @@ buildscript {
 
+class SecurityArgumentProvider implements CommandLineArgumentProvider {
+  @Internal
+  File commonSolrDir
```
just wondering, why is this one internal?

clayburn replied:
This one was difficult and may require a little more investigation. They are setting this property to the directory of the root project, which is almost certainly not an input to the test task in it's entirety. But the actual usage of this directory was not quite clear, something to do with the SecurityManager.

clayburn replied:
So in more detail, what happens here is that something generates a policy file for the SecurityManager which reads the `common-solr.dir` property and grants read access to that directory, which will evaluate to the root project directory.

It's pretty clear to me the intention isn't to treat the entire project as an input. This would include all source code, `.git`, `.gradle`, etc. So I think setting this to Internal is correct. I think, more ideally, this property would be more refined in scope to exactly what needs to be read by tests, but we lack that domain knowledge without more significant digging.

I'd be concerned that there may be inputs to the test task that are undeclared, given they need a permission like this. However, I don't think the change we are making here makes that situation any better or worse. We could say this about most any project's test suite.

runningcode commented on gradle/validation/jar-checks.gradle:
```diff
@@ -200,6 +200,10 @@ subprojects {
     dependsOn collectJarInfos
 
+    def outputFileName = "validateJarLicenses"
+    inputs.dir(file(project.rootDir.path + "/solr/licenses"))
```
```suggestion
    inputs.dir(file(project.rootDir.path + "/solr/licenses"))
```
is this not using absolue path sensitivity here? 
also let's give this input property a name

runningcode reviewed (commented):
thanks for the preview PR. added some comments.

clayburn commented on gradle/documentation/render-javadoc.gradle:
```diff
@@ -85,8 +86,8 @@ allprojects {
     offlineLinks += [
-        "https://docs.oracle.com/en/java/javase/11/docs/api/": javaJavadocPackages,
-        "https://junit.org/junit4/javadoc/4.12/": junitJavadocPackages
+        "https://docs.oracle.com/en/java/javase/11/docs/api/": project.relativePath(javaJavadocPackages),
```
Just stylistically, I think this would be a little cleaner to put the relative path logic in the definition on line 81 and 82 and leave these lines as is. My opinion though, so take it or leave it.

clayburn commented on gradle/documentation/render-javadoc.gradle:
```diff
@@ -49,7 +49,7 @@ allprojects {
       group "documentation"
 
-      taskResources = resources
+      taskResources = project.relativePath(resources)
```
It might make sense to move the relative path logic to the definition of `resources` on line 23. This way, in case someone authors a new task using `resources`, they will get the relative path automatically.

clayburn reviewed (commented):
I'd like to see, for each task we touched in some way, that we make that task [avoid configuration](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html). I wouldn't go through the whole project doing that at this time, just where we made changes.

tylerbertrand committed:
Fix spotless cache miss

Spotless custom steps are never up to date by default.
This is achieved using a randomly generated number attached to the custom step's state, which causes the custom step to never be up-to-date between runs.
Calling bumpThisNumberIfACustomStepChanges(int) pins this randomly generated number to the provided int, but it must be updated if a custom step changes, otherwise incorrect cache hits are possible.
